### PR TITLE
add storage_format parameter, allow for reading/writing data frames as CSVs / flat files

### PR DESCRIPTION
**Pull request by abelcastilloavant**
Extendable to other formats via implementation of `save_as` and `load_as` functions.

Example usage:
```
<R prompt> options("s3mpi.path" = "s3://my_favorite_s3_bucket")
<R prompt> df <- data.frame(a = 1, b = 2)
<R prompt> s3store(df, "test/write_as_csv", storage_format = "CSV", row.names = FALSE)
# writes the file as a CSV with no rownames
<R prompt> s3read("test/write_as_csv", storage_format = "CSV")
  a b
1 1 2
```

From bash:
```
<bash prompt> s3cmd get s3://my_favorite_s3_bucket/test/write_as_csv
<bash prompt> cat write_as_csv
"a","b"
1,2
```

### CODE REVIEW

File: R/s3.get.R [hunk 1]
```diff
@@ -71,11 +77,19 @@ s3.get <- function (path, bucket.location = "US", verbose = FALSE, debug = FALSE
...
   }
   ans
 }
+
+load_as_RDS <- function(filename, ...) {
+  readRDS(filename, ...)
+}
+
+load_as_CSV <- function(filename, ...) {
+  read.csv(filename, ...)
```
Comment (kirillseva) [inline R/s3.get.R]: stringsAsFactors = FALSE

File: R/s3.put.R [hunk 2]
```diff
@@ -54,3 +64,12 @@ run_system_put <- function(path, name, s3.cmd, check_exists, num_retries, backof
...
   }
 }
+
+save_as_RDS <- function(x, filename, ...) {
+  saveRDS(x, filename, ...)
+}
+
+
+save_as_CSV <- function(x, filename, ...) {
+  write.csv(x, filename, ...)
```
Comment (kirillseva) [inline R/s3.put.R]: row.names = FALSE (I think)

File: R/s3.get.R (hunk 1 above)
Comment (abelcastilloavant) [inline R/s3.get.R]: do we want that choice to live in this package or do we want the user to make that choice on their end?

**Comment by abelcastilloavant** [comment]
CI is failing, but tests pass locally:

```
> devtools::test("~/dev/s3mpi")
Loading s3mpi
Testing s3mpi
s3.get : .....
s3path :
s3read : .............
s3store : ...........

You rock!
```

**Comment by abelcastilloavant** [comment]
same CI failure as https://github.com/robertzk/s3mpi/pull/61

File: R/s3.get.R [hunk 3]
```diff
@@ -11,13 +11,18 @@
 #' @return For \code{s3.get}, the R object stored in RDS format on S3 in the \code{path}.
 #'    For \code{s3.put}, the system exit code from running the \code{s3cmd}
 #'    command line tool to perform the upload.
-s3.get <- function (path, bucket.location = "US", verbose = FALSE, debug = FALSE, cache = TRUE) {
+s3.get <- function (path, bucket.location = "US", verbose = FALSE, debug = FALSE, cache = TRUE, storage_format = "RDS", ...) {
+
+  if (!(storage_format %in% c("RDS", "CSV"))) {
```
Comment (robertzk) [inline R/s3.get.R]: R has a feature for this. ;)

```r
s3.get <- function(..., storage_format = c("RDS", "CSV"), ...) {
  storage_format <- match.arg()
```

File: R/s3.get.R [hunk 4]
```diff
@@ -35,7 +40,8 @@ s3.get <- function (path, bucket.location = "US", verbose = FALSE, debug = FALSE
     system2(s3cmd(), s3.cmd)
 
     ## And then read it back in RDS format.
-    readRDS(x.serialized)
+    load_from_file <- get(paste0("load_as_", storage_format), envir = as.environment("package:s3mpi"))
```
Comment (robertzk) [inline R/s3.get.R]: You shouldn't need `envir`.

File: R/s3.put.R [hunk 5]
```diff
@@ -11,7 +11,16 @@
 s3.put <- function (x, path, name, bucket.location = "US",
                     debug = FALSE, check_exists = TRUE,
                     num_retries = getOption("s3mpi.num_retries", 0), backoff = 2 ^ seq(2, num_retries + 1),
-                    max_backoff = 128) {
+                    max_backoff = 128, storage_format = "RDS", ...) {
+
+  if (!(storage_format %in% c("RDS", "CSV"))) {
```
Comment (robertzk) [inline R/s3.put.R]: Use `match.arg`

File: R/s3.put.R [hunk 6]
```diff
@@ -11,7 +11,16 @@
...
+    stop("The only storage formats supported at the moment are RDS and (for data frames) CSV.")
+  }
+
+  if (identical(storage_format, "CSV") && !is(x, "data.frame")) {
```
Comment (robertzk) [inline R/s3.put.R]: Use `==` if you have done argument matching / checking. Also use `is.data.frame`

File: R/s3.put.R (hunk 2 above)
Comment (robertzk) [inline R/s3.put.R]: Yes please.

File: R/s3.get.R (hunk 1 above)
Comment (robertzk) [inline R/s3.get.R]: @abelcastilloavant By default I think it should always be FALSE. That was a base library design mistake.

**Comment by abelcastilloavant** [comment]
Test driving locally:
```
> x <- data.frame(a = 1:5, b = 3:7)
> s3store(x, "abel/test_storage/CSV", storage_format = "CSV")
> s3store(x, "abel/test_storage/table", storage_format = "table", sep = "|", row.names = FALSE, quote = FALSE)
```
Contents of `test.csv`:
```
"","a","b"
"1",1,3
"2",2,4
"3",3,5
"4",4,6
"5",5,7
```

Contents of `test.txt`:
```
a|b
1|3
2|4
3|5
4|6
5|7
```

**Comment by robertzk** [comment]
@abelcastilloavant Use `row.names = FALSE` on `write.csv` for that reason.

**Comment by abelcastilloavant** [comment]
We can add that in to s3mpi, or we can expect the user to decide - what's your preference @robertzk?

**Comment by robertzk** [comment]
@abelcastilloavant I vote for adding it in.

**Pull request merged**